### PR TITLE
Retry downloads, respect URL list, validate tar hash

### PR DESCRIPTION
**Pull request by mtaufen**
I tested this with the instructions in the README: https://github.com/pjh/kubernetes/blob/gce-windows-cluster/cluster/gce/win1803/README-GCE-Windows-kube-up.md

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
1. Respects that NODE_BINARY_TAR_URL may be a list of URLs, instead of just one URL.
2. Improves network disruption tolerance by retrying downloads (similar to download-or-bust in Linux scripts).
3. Validates NODE_BINARY_TAR_HASH against downloaded files, if it is provided.

```release-note
NONE
```

@yujuhong @pjh

### CODE REVIEW

File: cluster/gce/win1803/common.psm1 [hunk 1]
```diff
@@ -86,5 +86,63 @@ function Get-InstanceMetadataValue {
...
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  $actual = Get-FileHash -Path $Path -Algorithm SHA1
+  if $actual -ne $Hash {
+    Log-Output "$Path corrupted, sha1 $actual doesn't match expected $Hash"
+    Throw ("$Path corrupted, sha1 $actual doesn't match expected $Hash")
+  }
+}
+
+# Attempts to downlaod the file from URLs, trying each URL in succession (with retries)
```
Comment (RA489) [inline cluster/gce/win1803/common.psm1]: s/downlaod/download

**Comment by mtaufen** [comment]
@pjh @yujuhong ready for review

File: cluster/gce/win1803/common.psm1 [hunk 2]
```diff
@@ -86,5 +86,69 @@ function Get-InstanceMetadataValue {
   }
 }
 
+function Validate-Hash {
```
Comment (pjh) [inline cluster/gce/win1803/common.psm1]: Add a comment that the hash is SHA1, or change the function name to indicate this.

File: cluster/gce/win1803/common.psm1 [hunk 3]
```diff
@@ -86,5 +86,69 @@ function Get-InstanceMetadataValue {
...
+  # to work.
+  if ($actual.Hash -ne $Hash) {
+    Log-Output "$Path corrupted, sha1 $actual doesn't match expected $Hash"
+    Throw ("$Path corrupted, sha1 $actual doesn't match expected $Hash")
+  }
+}
+
+# Attempts to download the file from URLs, trying each URL in succession (with retries)
+# until it succeeds. It will loop through the URLs list forever until it has a success.
+# If successful, it will write the file to OutDir, with a
```
Comment (pjh) [inline cluster/gce/win1803/common.psm1]: OutFile

File: cluster/gce/win1803/common.psm1 [hunk 4]
```diff
@@ -86,5 +86,69 @@ function Get-InstanceMetadataValue {
...
+      } Catch {
+        $message = $_.Exception.ToString()
+        Log-Output "Failed to download file from $url. Will retry. Error: $message"
+        continue
+      }
+      # Attempt to validate the hash
+      if ($Hash -ne $null) {
+        Try {
+            Validate-Hash -Hash $Hash -Path $OutFile
+          } Catch {
```
Comment (pjh) [inline cluster/gce/win1803/common.psm1]: There's a couple extra leading spaces here.

File: cluster/gce/win1803/k8s-node-setup.psm1 [hunk 5]
```diff
@@ -318,14 +318,13 @@ function DownloadAndInstall-KubernetesBinaries {
...
-  Invoke-WebRequest $uri -OutFile ${tmp_dir}\${filename}
-
-  # TODO(yujuhong): Verify hash of the tarball.
+  $urls = ${kube_env}['NODE_BINARY_TAR_URL'].Split(",")
+  $filename = Split-Path -leaf $urls[0]
+  $hash = $null
+  if ($kube_env.ContainsKey('NODE_BINARY_TAR_HASH')) {
+    $hash = ${kube_env}['NODE_BINARY_TAR_HASH']
+  }
+  MustDownload-File -Hash $hash -OutFile ${tmp_dir}\${filename} -URLs $urls
```
Comment (pjh) [inline cluster/gce/win1803/k8s-node-setup.psm1]: Should we change other Invoke-WebRequest calls in this file to use MustDownload-File?

File: cluster/gce/win1803/common.psm1 [hunk 6]
```diff
@@ -86,5 +86,69 @@ function Get-InstanceMetadataValue {
...
+  if ($actual.Hash -ne $Hash) {
+    Log-Output "$Path corrupted, sha1 $actual doesn't match expected $Hash"
+    Throw ("$Path corrupted, sha1 $actual doesn't match expected $Hash")
+  }
+}
+
+# Attempts to download the file from URLs, trying each URL in succession (with retries)
+# until it succeeds. It will loop through the URLs list forever until it has a success.
+# If successful, it will write the file to OutDir, with a
+# filename matching the base (final) segment of the URL. You can optionall provide a Hash
```
Comment (pjh) [inline cluster/gce/win1803/common.psm1]: optionally

File: cluster/gce/win1803/common.psm1 [hunk 7]
```diff
@@ -86,5 +86,69 @@ function Get-InstanceMetadataValue {
...
+function MustDownload-File {
+  param (
+    [parameter(Mandatory=$false)] [string]$Hash,
+    [parameter(Mandatory=$true)] [string]$OutFile,
+    [parameter(Mandatory=$true)] [System.Collections.Generic.List[String]]$URLs
+  )
+
+  # TODO(mtaufen): We could probably set this in a more general place.
+  # Disable progress bar to increase download speed.
+  $ProgressPreference = 'SilentlyContinue'
```
Comment (pjh) [inline cluster/gce/win1803/common.psm1]: You could move this to Set-PrerequisiteOptions in k8s-node-setup.psm1, or (perhaps better) just move this to be "global" at the top of this file.

**Review by pjh — COMMENTED** [review]
This looks great, thanks!

File: cluster/gce/win1803/k8s-node-setup.psm1 (hunk 5 above)
Comment (mtaufen) [reply cluster/gce/win1803/k8s-node-setup.psm1]: Good point. Done.

File: cluster/gce/win1803/common.psm1 (hunk 6 above)
Comment (mtaufen) [reply cluster/gce/win1803/common.psm1]: Done.

File: cluster/gce/win1803/common.psm1 (hunk 7 above)
Comment (mtaufen) [reply cluster/gce/win1803/common.psm1]: Moved to top of file.

File: cluster/gce/win1803/common.psm1 (hunk 4 above)
Comment (mtaufen) [reply cluster/gce/win1803/common.psm1]: Done.

File: cluster/gce/win1803/common.psm1 (hunk 3 above)
Comment (mtaufen) [reply cluster/gce/win1803/common.psm1]: Done.

File: cluster/gce/win1803/common.psm1 (hunk 2 above)
Comment (mtaufen) [reply cluster/gce/win1803/common.psm1]: Done.

**Comment by mtaufen** [comment]
Looks like this had a small bug, fix in https://github.com/kubernetes/kubernetes/pull/74324

**Pull request merged**